### ISSUE TITLE
Repack with better compression

### ISSUE THREAD

**Issue by maxim-zhao** [maxim-zhao/STH1EDWV#10]
psgcompr and/or LZSA1 seem likely to make things a bit smaller and therefore pack better, while also potentially speeding things up a bit.

https://github.com/maxim-zhao/bmp2tilecompressors/raw/master/benchmark.png?raw=true

**Comment by iamgreaser** [comment]
Apologies if I'm telling you nothing you already know.

For the level layouts, the relevant function is at $00A10 in the ROM and is 48 bytes long, of which the first 3 bytes sets the destination address. I've actually managed to fit a home-baked LZSS decompressor into that space with a few bytes to spare, and that managed to free up a bit more than 8 KB of ROM space, which is more than the amount of free space available if you ignore unused assets (I believe it's somewhere around 1.5 KB).

But even with the current scheme it's not hard to cut the size of the function down - it checks in 3 different places to see if the input length has been reached, and it tracks the previous RLE value in an (IY+d) value which gets mentioned in 3 different ops. So you've got room.

LZSA1... haven't heard of it, but:

> As an example of LZSA1's simplicity, a size-optimized decompressor on Z80 has been implemented in 67 bytes.

I believe my LZSS-based decompressor does it in 41 bytes, even if the compression isn't quite as tasty. But you can probably find a spare 22 bytes somewhere if you're willing to steal them from some other functions... and actually I did manage to pull that off, the one place that calls the unpack function is at $02243 in the ROM and I somehow managed to eke out an extra 26 bytes over the $2212-$2245 range. So, it shouldn't be too drastic to fit it in as long as you can split the decompressor sanely.

I've yet to look into repacking the art. I can think of some ways to not only speed up decompression but also potentially pack it a bit better.

For that, the function is at $00405 in the ROM and is 252 bytes long including the 8-byte lookup table. You can speed this one up and shrink it down *easily*, it loads the next bit from the selection bit stream in a *horribly* inefficient way. It *is* called in a lot of places though.

The Phantasy Star Gaiden decompressor could probably slot into that space as-is, the art tile decompressor really is that wasteful.

Immediately following that is a loader for RLE-compressed tilemap data. $00501 in the ROM, 101 bytes. Also called in a lot of places.

Of course, most of the art tiles and art tilemaps have hardcoded pointers so you're going to be searching the entire codebase for the calls to those functions.

... I'll just save you the pain of hunting down all the constant calls if you haven't already. Here they are. It covers everything except for the two calls that grab the 2 art tile pointers from the current level header.

Sorry about the weird format, I'm currently working on a disassembly of my own and so far it seems the first 3 banks are position-independent. Basically, xx:yyyy maps to physical address `(xx<<14)|(yy&$3FFF)`, `banksetting` indicates that at the given slot we have the given bank mapped over the given range (start <= x < end), and we're also pretending that slot 3 exists and that RAM is at bank $F0. But I have reason to believe this is literally every single pointer of these two types in the ROM that isn't in the level headers.

Art tiles here tend to cover a single `LD HL, $xxxx` op. Art tilemaps cover more than just one op, so you'll have to look at the ranges to see what actually needs patching.

```
label 09:2000 byte ART_09_2000
   banksetting 0 09 00:1296 00:1299
label 09:351F byte ART_09_351F
   banksetting 0 09 00:1411 00:1414
   banksetting 0 09 00:1580 00:1583
label 0A:4294 byte ART_09_4294
   banksetting 1 0A 01:5F2D 01:5F30
label 0A:4B0A byte ART_09_4B0A
   banksetting 1 0A 00:12A1 00:12A4
   banksetting 1 0A 00:26B6 00:26B9
label 0A:526B byte ART_09_562B
   banksetting 1 0A 00:0C94 00:0C97
label 0A:5942 byte ART_09_5942
   banksetting 1 0A 00:0CF6 00:0CF9
label 0B:AEB1 byte ART_09_AEB1
   banksetting 2 0B 01:7031 01:7034
   banksetting 2 0B 02:8074 02:8077
label 0B:B92E byte ART_09_B92E
   banksetting 2 0B 00:0C9F 00:0CA2
   banksetting 2 0B 00:0D01 00:0D04
   banksetting 2 0B 00:1575 00:1578
   banksetting 2 0B 00:2172 00:2175
label 0C:0000 byte ART_0C_0000
   banksetting 0 0C 00:0C89 00:0C8C
   banksetting 0 0C 00:25A9 00:25AC
label 0C:1801 byte ART_0C_1801
   banksetting 0 0C 00:0CEB 00:0CEE
   banksetting 0 0C 00:26AB 00:26AE
label 0F:DA28 byte ART_0C_DA28
   banksetting 3 0F 01:7916 01:7919
label 0F:E508 byte ART_0C_E508
   banksetting 3 0F 02:84BC 02:84BF
   banksetting 3 0F 02:9291 02:9294
label 0F:EF3F byte ART_0C_EF3F
   banksetting 3 0F 02:A816 02:A819
   banksetting 3 0F 02:BB94 02:BB97

label 05:6000 byte ARTMAP_05_6000
   banksetting 1 05 00:12B4 00:12C5
label 05:612E byte ARTMAP_05_612E
label 05:61E9 byte ARTMAP_05_61E9
   banksetting 1 05 00:1593 00:15B3
label 05:627E byte ARTMAP_05_627E
label 05:63F6 byte ARTMAP_05_63F6
   banksetting 1 05 00:0CB2 00:0CD4
label 05:653B byte ARTMAP_05_653B
label 05:66AB byte ARTMAP_05_66AB
   banksetting 1 05 00:0D14 00:0D36
label 05:67FE byte ARTMAP_05_67FE
   banksetting 1 05 00:1424 00:1435
label 05:6830 byte ARTMAP_05_6830
   banksetting 1 05 00:25BC 00:25CC
label 05:69A9 byte ARTMAP_05_69A9
   banksetting 1 05 00:267D 00:268D
label 05:6C61 byte ARTMAP_05_6C61
   banksetting 1 05 00:26C9 00:26D9
```

**Comment by maxim-zhao** [comment]
Yeah, I do know that already :) I have the existing disassemblies plus a custom hacked together version I need to unbreak and push to the disassembly dirctory.

LZSS is a general purpose compressor for Z80 platforms, and if using it there's a decent chance we'll have so much free space for extra code we won't know what to do with it. The main blocker for me is that (1) I haven't been spending time on this project much recently and (2) I didn't get around to biting the bullet on the concept of "optional code changes with associated linker fixups" to inject things like compressors, cheats, splash screens, early ending hacks, changing the data LUTs to be bank+offset instead of just 16-bit offsets, etc.

It's this second part that's probably more tricky - the data placement and linkage code I wrote took a while and a lot of mental energy, and I'd really need to get back into the code. I did start on an unfinished attempt at refactoring the internal data model which was supposed to help, but now I look at it and I can't remember how...

I think the end result is that I should either emit assembly code and call into a real assembler + linker, or I need to write a multi-pass linker to deal with the multi-tiered parts of the build process. Both seem like hard work.
